### ISSUE TITLE
New Gamemode: 2 Player Full Game

### ISSUE THREAD

**Issue by WilliamWeberBerrutti** [opentyrian/opentyrian#22]
I've had an idea for a new gamemode called 2 Player Full Game, which is essentially the 1 Player Full Game but for 2 players. That's one thing I missed since playing the original Tyrian: to allow 2 players with full game experience. The store could have all items of the 1 Player Full Game, but maybe with some modifications:

- reduce by half all prices (necessary?)
- each player has their items separated OR MAYBE P1 has all except Rear Guns and Sidekicks, and P2 has all except Front Guns OR MAYBE both players have both Front and Rear Guns (would it cause too much spam?)
- it'll be necessary to define ships P1 and P2 can buy. Maybe create 2 more ships based on the 2 Player Arcade mode to be the most powerful ones, with the highest armor stats? Or simply remove the ship submenu from the store?

**Comment by jotaf98** [comment]
That would be amazing, but likely very hard to implement with all those changes!

Even without thinking about balance/fairness, just directly merging Story Mode and 2 Players Arcade opens many questions for implementation/GUI:
- How is the shop handled in the GUI? It only shows one ship at a time. One possibility could be to have 2 "Upgrade ship" options in the main menu: "Upgrade ship 1" and "Upgrade ship 2". And then somehow hackily swap their state in the code.
- How is the generator status for both ships displayed? This takes up considerable space, and is missing in the arcade version.

Suffice to say, something that looks like simple moving around of existing data would be much easier to implement than whole new GUIs.

**Comment by WilliamWeberBerrutti** [comment]
> How is the shop handled in the GUI? It only shows one ship at a time. One possibility could be to have 2 "Upgrade ship" options in the main menu: "Upgrade ship 1" and "Upgrade ship 2". And then somehow hackily swap their state in the code.

That's a nice idea! Also the ship on the left could be drawn only after clicking on those new options.


> How is the generator status for both ships displayed? This takes up considerable space, and is missing in the arcade version.

Maybe it could be used those parts I highlighted in red, creating new sections.

![Screenshot_20211117_172204_edited](https://user-images.githubusercontent.com/7718402/142277030-1d9c0f3a-91a2-46a2-ba32-dc9d5ff981f0.png)

or

![Screenshot_20211117_172204_edited2](https://user-images.githubusercontent.com/7718402/142277039-a2ae9b34-1797-41ad-9f11-8badcf7aadbc.png)

The design could be based on the red horizontal bar of the player level, but recoloring maybe (to blue?), or recolor the player level bar to something else and the generator energy one to red?

**Comment by seifer7** [comment]
Just letting you guys know I am working on a 2 Player Full Game Mode.
It's going well. I will update you when I have something ready for you to try out!

**Comment by jotaf98** [comment]
That's awesome!

**Comment by seifer7** [comment]
Note that this will work with this original project without breaking any existing functions and hopefully will be approved and merged when I put in a pull request. :D

Here are some sneak peeks
![image](https://user-images.githubusercontent.com/13601073/211465412-b73098c1-0b1f-4151-8edf-6cbad9779410.png)
![image](https://user-images.githubusercontent.com/13601073/211465458-cdab86d2-4a23-4952-bfd7-a07074911678.png)
![animated sneak peek](https://user-images.githubusercontent.com/13601073/211467489-bbd512d0-3da4-4468-81ba-c13c5882ab3e.gif)

**Comment by carlreinke** [comment]
Sorry to rain on your parade here, but a new game mode is outside the scope of what OpenTyrian is intended to be (which boils down the "Tyrian 2.1 with quality of life improvements").  That doesn't mean you shouldn't make it, of course, but you shouldn't expect that it will be merged into this repo.

**Comment by WilliamWeberBerrutti** [comment]
@seifer7 even if your work isn't merged, I'd have interest on playing this new gamemode (and probably some people I know too). Maybe you can begin a new OpenTyrian variant with not only QoL improvements but new ones, not present in the original game, like this new gamemode you developed.

**Comment by jotaf98** [comment]
@seifer7 I'd be interested as well, you can probably fork and just have a slightly different title to make it obvious, like "OpenTyrian with couch co-op" or similar!

@carlreinke Although it's described as a new game mode, in gameplay terms it's just a direct combination of existing modes.

There are 4 possible combinations of the binary attributes [story mode, arcade mode] and [single player, multiplayer]. Tyrian included 3 of the 4 combinations, leaving out only story mode + multiplayer.

There doesn't seem to have been a reason to leave such a straightforward combination of existing functionality, other than lack of dev resources. So it could be argued to be a QOL improvement if you want.

**Comment by jotaf98** [comment]
Since what kept me from trying this mode before was compiling it for Windows, in case anyone else is in the same situation, you'll find a compiled binary here (place it in the same folder as the original executable):
https://drive.google.com/file/d/13YpqMoogJEU7frw6LJ5oLt6-WQPKu0km/view?usp=sharing

Have fun!

**Comment by WilliamWeberBerrutti** [comment]
Maybe it should be good for both players to start with 20k money, and maybe give 2x points for kills so it could be properly balanced.

**Comment by seifer7** [comment]
@jotaf98 @WilliamWeberBerrutti Thanks for your interest guys!

I have enabled "Issues" on my fork over at https://github.com/seifer7/tyrianreloaded/tree/two-player-full-mode

Feel free to post the bugs or suggestions over there!
